src/app/CommandSender.h

```diff
@@ -53,76 +53,143 @@ namespace app {
+         * errors will be delivered through this call in the aError field:
+         *
+         * - CHIP_ERROR_TIMEOUT: A response was not received within the expected response timeout.
+         * - CHIP_ERROR_*TLV*: A malformed, non-compliant response was received from the server.
+         * - CHIP_ERROR encapsulating a StatusIB: If we got a non-path-specific
+         *   status response from the server.  In that case,
+         *   StatusIB::InitFromChipError can be used to extract the status.
+         * - CHIP_ERROR encapsulating a StatusIB: If we got a path-specific
+         *   status response from the server.  In that case,
+         *   StatusIB::InitFromChipError can be used to extract the status.
+         *      - Note path specific error only come here happens if UsingExtendedPathCallbacks()
+         *        returns false.
+         *      - There isn't a guaranteeded way to differentiate between a non-path-specific error and a
+         *        path-specific error from the OnError callback.
+         * - CHIP_ERROR*: All other cases.
+         *
+         * The CommandSender object MUST continue to exist after this call is completed. The application shall wait until it
+         * receives an OnDone call to destroy and free the object.
+         *
+         * @param[in] apCommandSender The command sender object that initiated the command transaction.
+         * @param[in] aError          A system error code that conveys the overall error code.
+         */
+        virtual void OnError(const CommandSender * apCommandSender, CHIP_ERROR aError) {}
+
+        /**
+         * OnDone will be called when CommandSender has finished all work and is safe to destroy and free the
+         * allocated CommandSender object.
+         *
+         * This function will:
+         *      - Always be called exactly *once* for a given CommandSender instance.
+         *      - Be called even in error circumstances.
+         *      - Only be called after a successful call to SendCommandRequest returns, if SendCommandRequest is used.
+         *      - Always be called before a successful return from SendGroupCommandRequest, if SendGroupCommandRequest is used.
+         *
+         * This function must be implemented to destroy the CommandSender object.
+         *
+         * @param[in] apCommandSender   The command sender object of the terminated invoke command transaction.
+         */
+        virtual void OnDone(CommandSender * apCommandSender) = 0;
+    };
+
+    // CommandSender::ExtendedCallback::OnResponse is public SDK API, so we cannot break
+    // source compatibility for it. To allow for additional values to be added at a future
+    // time without constantly changing the function's declaration parameter list, we are
+    // defining the struct AdditionalResponseData and adding that to the parameter list
+    // to allow for future extendability.
+    struct AdditionalResponseData
+    {
+        Optional<uint16_t> mCommandRef;
```

Comment:
Suggest "command_ref"/"commandRef" rather than mCommandRef. Public fields of structs should not look like private members. Same in ErrorData.